src/shared.rs

```diff
@@ -0,0 +1,148 @@
+    pub fn as_deref(&self) -> ModifierSet<&str> {
+        ModifierSet(&self.0)
+    }
+
+    /// Construct a modifier set from a string,
+    /// where modifiers are separated by the character `.`.
+    ///
+    /// It is not unsafe to use this function wrongly, but it can produce
+    /// unexpected results down the line. Correct usage should ensure that
+    /// `s` does not contain any empty modifiers (i.e. the sequence `..`)
+    /// and that no modifier occurs twice.
+    pub fn new_unchecked(s: S) -> Self {
+        Self(s)
+    }
+
+    /// Construct an empty modifier set.
+    pub fn empty() -> Self
+    where
+        S: Default,
+    {
+        Self(S::default())
+    }
+
+    /// Whether `self` is empty.
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    /// Add a modifier to the set, without checking that it is a valid modifier.
+    ///
+    /// It is not unsafe to use this method wrongly, but that can produce
+    /// unexpected results down the line. Correct usage should ensure that
+    /// `modifier` is not empty and doesn't contain the character `.`.
+    pub fn add_unchecked(&mut self, m: &str)
+    where
+        S: for<'a> AddAssign<&'a str>,
+    {
+        if !self.0.is_empty() {
+            self.0 += ".";
+        }
+        self.0 += m;
+    }
+
+    /// Iterate over the list of modifiers in an arbitrary order.
+    pub fn iter(&self) -> impl Iterator<Item = &str> {
+        self.0.split('.').filter(|s| !s.is_empty())
+    }
+
+    /// Whether the set contains the modifier `m`.
+    pub fn contains(&self, m: &str) -> bool {
+        self.iter().any(|lhs| lhs == m)
+    }
+
+    /// Whether all modifiers in `self` are also present in `other`.
+    pub fn is_subset(&self, other: ModifierSet<&str>) -> bool {
+        self.iter().all(|m| other.contains(m))
+    }
+
+    /// Find the best match from the list.
+    ///
+    /// To be considered a match, the modifier set must be a superset of
+    /// (or equal to) `self`. Among different matches, the best one is selected
+    /// by the following two criteria (in order):
+    /// 1. Number of modifiers in common with `self` (more is better).
+    /// 2. Total number of modifiers (fewer is better).
+    pub fn best_match_in<'a, T>(
+        &self,
+        variants: impl Iterator<Item = (ModifierSet<&'a str>, T)>,
+    ) -> Option<T> {
+        let mut best = None;
+        let mut best_score = None;
+
+        // Find the best table entry with this name.
+        for candidate in variants.filter(|(set, _)| self.is_subset(*set)) {
+            let mut matching = 0;
+            let mut total = 0;
+            for modifier in candidate.0.iter() {
+                if self.contains(modifier) {
+                    matching += 1;
+                }
+                total += 1;
+            }
+
+            let score = (matching, core::cmp::Reverse(total));
+            if best_score.map_or(true, |b| score > b) {
```

Comment:
```suggestion
            if best_score.is_none_or(|b| score > b) {
```

CI uses stable Rust so I think this should be fine.

Reply:
Yep, I wrote this before the recent MSRV bump.

Reply:
Codex CI is separate, and used latest stable Rust as of now.

Reply:
ah, the more you know